src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -3375,23 +3375,35 @@ BackgroundTaskReadyToRun(BackgroundTask *task)
 
 
 /*
- * GetRunnableBackgroundTask returns the first candidate for a task to be run. When a task
- * is returned it has been checked for all the preconditions to hold.
+ * GetNextReadyBackgroundTask returns a cancelling, runnable or running task to be
+ * run by task queue monitor. It will try to find the next task in round-robin fashion
+ * so that tasks will not starve to wait for their excecution. When a task is returned,
+ * it has been checked for all the preconditions to hold.
  *
  * That means, if there is no task returned the background worker should close and let the
  * maintenance daemon start a new background tasks queue monitor once task become
  * available.
  */
 BackgroundTask *
-GetRunnableBackgroundTask(void)
+GetNextReadyBackgroundTask(int64 lastMonitoredTaskId)
 {
 	Relation pgDistBackgroundTasks =
 		table_open(DistBackgroundTaskRelationId(), ExclusiveLock);
 
+	/*
+	 * Ordering of status is important.
+	 * 1) We should change Cancelling status to Cancelled,
+	 * 2) We should then select runnable or running tasks.
+	 */
 	BackgroundTaskStatus taskStatus[] = {
-		BACKGROUND_TASK_STATUS_RUNNABLE
+		BACKGROUND_TASK_STATUS_CANCELLING,
+		BACKGROUND_TASK_STATUS_RUNNABLE,
+		BACKGROUND_TASK_STATUS_RUNNING
 	};
```

Comment:
I am surprised by this change, wouldn't any task that is ready to be started at this stage be marked as `RUNNABLE`?

To me a task that is `CANCELLING` should not be touched, because the backend of the task itself is getting a signal. This should trigger the backend to stop execution and the state should transition not when we are looking for a next task, but instead when the stopped backend is inspected and the associated task updated.

Same would de for `RUNNING` these tasks are not ready to start because they already should be started.

I didn't look deeply into the rest of the change, but this part stood out to me while skimming the change.

Either we are doing funky things here, or the comment above is very misleading. And doesn't explain why we are iterating `CANCELLING` and `RUNNING` tasks.

Reply:
You are right. Here some updates. I refactored the code such that queue monitor looks for runnable tasks and updates their states in FSM loop. It looks for task table to fetch runnable tasks and also has a hash table to store currently running task and executor pairs. Cancelling is also updated in the FSM now. I made FSM states little bit more readable.